Add sidechain registration and recovery page in Lisk docs - Closes #1517 (#1632)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/understand-blockchain/interoperability/sidechain-registration-and-recovery.adoc

= Sidechain Registration & Recovery
Chris Braithwaite <christopher.braithwaite@lightcurve.io> Mona Bärenfänger <mona@lightcurve.io>
:description: How to register, terminate, and recover a sidechain.
:toc:
:idprefix:
:idseparator: -
:docs-sdk: lisk-sdk::
:imagesdir: ../../assets/images

:url_ccm: understand-blockchain/interoperability/communication.adoc
:url_ccu: understand-blockchain/interoperability/communication.adoc#-relaying-ccms-in-a-ccu
:url_ccm1: understand-blockchain/interoperability/communication.adoc#inducing_state_changes_across_chains_with_ccms
:url_bls: pages/glossary.adoc#bls
:url_bft_weights: pages/glossary.adoc#bft-weight
:url_outbox_rootwitness: understand-blockchain/interoperability/communication.adoc
:url_sdk_token_module: {docs_sdk}modules/token-module.adoc
:url_custom_module: understand-blockchain/sdk/modules-commands
:url_nft_module: tutorial/nft.adoc#the-nft-module
:url_lip: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0045.md

//:url_how_blockchain_works: intro/how-blockchain-works.adoc
//:url_transactions: understand-blockchain/lisk-protocol/transactions.adoc

> == Overview - What is a Sidechain?

Author: Ah yes, I like your version Andreas... OK, I have now used it so then please ignore the other changes mentioned regarding this section. Many thanks 👍

> A sidechain is actually a separate blockchain that is attached to its mainchain using a two-way peg.

Author: Many thanks Andreas. I have modified  this sentence as you suggested 👍

Sidechains are aligned in parallel with the main blockchain.
Sidechains can facilitate moving digital assets from their origin, namely the original blockchain to a separate blockchain which is known as a sidechain.
Furthermore, it is possible to move these assets back to the original blockchain if necessary, resulting in being able to securely interchange data and assets.

> Simply put, a sidechain will allow nodes containing assets or data from the parent blockchain to be securely used in a separate blockchain (the sidechain), and hence, these assets and data can then be transferred back to the original blockchain.

Author: As recommended, I have removed these 2 lines and reworded the last sentence as suggested 👍

> In turn, the advent of sidechains will further increase the performance of the parent blockchain by taking some of the workload, and furthermore, this increases the level of security as all transactions that are executed, are secured by the participants (generators/forgers), of the mainchain.

Author: Many thanks. Now adapted accordingly 👍

> image:understand-blockchain/interop/sidechain.png[, align="center" ,500]

Author: I was just trying to give a high-level overview of the basic concept. However, I think you are right Andreas, we do not actually need it here as the main topics are the registration and recovery of a sidechain. Therefore, imho, by the time the reader is at this point in the docs, they should have some knowledge of the principle of sidechains. Therefore I have removed the image 👍

Figure 1: Sidechain overview.

> == Sidechains in Lisk

Reviewer: @chris529 If you don't want to use my suggestion, it's totally fine. But then I have to list all individual errors in this section which I haven't done yet.

> == Sidechains in Lisk

Author: @AndreasKendziorra Many thanks, I like your suggestion very much, it is excellent and I plan to use it. I will try to get to it today 👍

> == Sidechains in Lisk

Author: I have now updated this section with your recommendation Andreas, many thanks 👍

== Life Cycle of a Sidechain

The life cycle of a sidechain can be broken down into 3 parts, corresponding to the following 3 values of the account status property: ''registered'', ''active'', and ''terminated''.

Firstly, a sidechain registers on the mainchain with a *sidechain registration command*.
This command creates the sidechain account on the mainchain.
Specifically, it creates an entry in the chain data substore, with the initial status set to 'registered'; an entry in the channel data substore, with an empty inbox and outbox; coupled with an entry in the validators data substore, holding the initial sidechain validators set.
Thereafter, the mainchain account is similarly created on a sidechain with a *mainchain registration command*.

After a sidechain has been registered on the mainchain, it cannot receive any cross-chain message and does not need to follow the liveness rule, until the first sidechain CCU has been included in the mainchain.
At this point, the status of the sidechain account on the mainchain is updated to ''active'' and the liveness requirement rule is then enforced.

> (The liveness rule requires active sidechains to prove their liveness once every 30 days, or the sidechain account is terminated).

Author: Good point...I have not committed this change, as I have updated it accordingly, and moved it 1 line up as suggested 👍

Therefore, if no CCU is received within 30 days, the *chain account is terminated* and no more CCMs can be sent to or received from the sidechain.
A sidechain account can also be terminated if the sidechain posts a CCU containing a CCM with an invalid schema or with an invalid sending chain ID.
A message targeting a terminated chain is bounced on the mainchain instead of being forwarded.
When this happens, a "terminated sidechain" message is emitted by the mainchain, targeting the original sending chain.
When this message is processed, the chain is also terminated in the sending chain, blocking future messages.

When a chain is terminated, a "terminated state" account is created, storing the last certified state root of the terminated chain.
Then, users can recover tokens from the terminated sidechain with a <<State Recovery Command>> or proceed with the <<Message Recovery Initialization Command>> to kick off the recovery of pending cross-chain messages.
The sidechain termination and recovery process is covered later on this page.

> == Sidechain Registration

Author: OK, I can see your point, it is maybe a bit repetitive, so I have removed the "Sidechain Registration" section as suggested.

Firstly sidechain registration is required to ultimately ensure interoperability with a separate chain can be achieved with the Lisk mainchain.
In effect, this is managed by a registration command that creates a sidechain account within the Lisk mainchain.
This can be thought of as the initial step, being the creation and initiation of the communication channel that is required for two separate, different chains to interoperate and function seamlessly together.

> To bring this to fruition a cross-channel chain protocol needs to be established, which is the chain registration process defining the necessary protocol rules and data structures.

Author: This paragraph has been previously removed 👍

> At this point, it should be noted that there are two registration commands which are utilized depending on whether it is the Lisk mainchain or the sidechain we are dealing with.

Author: Done 👍

To explain this briefly, for the Lisk mainchain the sidechain registration command is used.

> From the sidechain perspective, this is performed by the mainchain registration command, both of these are covered further in this section.

Author: Yes, this was part of the previously removed section.

== How to register a sidechain

As mentioned in the previous paragraph to enable a sidechain to operate, it first has to be registered in the Lisk mainchain via a sidechain registration command.
This requires that the sidechain registration command is processed, which will then result in the sidechain account status being set to registered.

> However, at this point, the cross-channel is not active, so no cross-chain messages (CCMs) can be sent yet.

Author: Cool...Now updated and replaced 'cross-channel" with "channel"

> For the processing of this command, it signifies that a sidechain account in the mainchain state exists, and is associated with a name and a unique network identifier.

Author: Your English is excellent Andreas!  In fact, I am not sure we even need to mention processing the command in this manner here, so how about we reword it to something like the following..?

"Hence, a sidechain account that is already associated with a name and a unique network identifier, exists in the mainchain state. Therefore, each new sidechain occupies a certain namespace in the ecosystem".

Therefore, each new sidechain occupies a certain namespace in the ecosystem.

> ==== Registration Message

Author: Yes, I think it fits better there...Now done 👍

The role of the registration message is to allow for a safe activation of the channel between the sending and receiving chains.
It guarantees that when the CCU activating the channel is executed, a registration transaction has been included on the partner chain.
When a sidechain is registered on the mainchain, as previously mentioned, an ecosystem wide chain ID and name are assigned to this chain.
The chain name, network ID, and the token used for the message fees are included in a registration message that is appended to the sidechain outbox.
When the first cross-chain update containing messages is sent to the sidechain, the equality between the properties in the registration message and the ones in the interoperability store is verified.

==== Sidechain Registration Process

> Firstly, for a sidechain to be interoperable, it has to be registered in the Lisk mainchain via a <<Sidechain Registration Command>>.

Author: Ha Ha...Yes, you are right!! I think I was maybe trying to over-emphasize this point! :-) OK, I have removed it, and slightly adapted the first sentence accordingly 👍

A transaction with this command can be sent by any user account in the Lisk Mainchain that possesses adequate funds to pay the required fee.
The processing of this command signifies the creation of a sidechain account in the mainchain state that is associated with a unique network identifier and a name.

> Therefore, each new sidechain occupies a certain namespace in the ecosystem.

Author: Well spotted...I have now removed them 👍

Additionally, every newly registered sidechain can increase the size of every cross-chain update command, (CCU) posted on the mainchain (due to the increasing size of the xref:{url_outbox_rootwitness}[outboxRootWitness] property of the command).
For these two reasons, the minimum fee for this command has an added constant similar to the extra fee in a *delegate registration command*.

> The value of this extra registration fee is `REGISTRATION_FEE` is LSK.

Author: Fair point...I have now removed it 👍

Once the sidechain registration command is processed, the sidechain account status is set to registered.
In this state, the cross-chain channel is still not active, so the users on the mainchain or other chains cannot send xref:{url_ccm1}[cross-chain messages] (CCMs) to this sidechain yet. Moreover, the *liveness condition* to maintain the channel is not enforced, this means that there is no specific time requirement for a sidechain to be activated on the mainchain, it can stay in the registered status for any period of time.

> When a first valid cross-chain update command from this sidechain is processed, the sidechain status is changed to active, making it active in the ecosystem. Now it is possible to send CCMs to the sidechain and the liveness condition is enforced.

Author: I presume you mean the majority of this section? I have tried to reword it in a more high-level and concise manner, just to maintain the flow here..

Once a sidechain is terminated, it is not possible to send or receive cross-chain messages anymore.
Therefore, the ability of a sidechain to interoperate with other chains can be revoked, i.e., terminated, permanently.
In effect, the sidechain is disconnected from the rest of the ecosystem, and any assets (e.g. fungible or non-fungible tokens), cannot be moved either to or from the respective sidechain.
Therefore, this means that users will not be able to send any assets that they originally held on the sidechain, back to the original native chain, in other words, the chain where the assets were initially created.

The following section covers the recovery mechanism deployed by Lisk to address this problem, which will noticeably improve the user experience of the Lisk ecosystem without affecting the security guarantees of the general interoperability solution.
Using dedicated commands that are part of the Interoperability module, users can recover their assets directly on the native chain.
Furthermore, pending messages that were stuck unprocessed in the outbox of the terminated chain can be recovered as well, and their effect reverted on the sending chain.

> Sidechain termination specifically occurs when the sidechain in question has been inactive for too long.

Author: Thanks Sergey, I have now added it in as suggested 👍

For example, this could be caused by not posting a transaction with a cross-chain update (CCU), command for more than 30 days, or in the case whereby one was posted with a malicious CCU command on the mainchain.
Hence, it is useful to provide a trustless on-chain mechanism to recover tokens, messages, and information from terminated sidechains.

=== Termination causes

A sidechain can be terminated for various reasons as described below in the following paragraphs.

* Liveness requirement

As previously mentioned, active sidechains are required to prove they are still live and functioning by posting a CCU at least once within a 30-day time period.

> Therefore violating the liveness requirement (only on the mainchain), result in terminating the sidechain.

Author: Many thanks...Now added 👍

=== State recovery from the Sidechain Root

This mechanism has been enabled in order to ensure that it is possible to recover a specific entry from a substore (i.e. the collection of key-value pairs with a common store prefix), of a module store of a terminated sidechain.
In this context, the term "recover" is referring to triggering a specific state transition defined as part of the relevant module protocol logic.
Specifically, it is based on the sidechain state root, `stateRoot`, set in the last certificate before sidechain termination.

In the context of the mainchain, a valid state recovery command can recover the assets that were held in the terminated sidechain.
In the context of a sidechain, it can recover an entry in a recoverable module store from a terminated sidechain.
A recoverable module is classed as any module that exposes a recover function, which includes the Token module (for any custom token), and the xref:{url_nft_module}[NFT module].
This recovery mechanism requires the initialization commands.

==== State Recovery Initialization Command

> This command has two purposes.

Author: As discussed earlier in Slack, it has now been updated 👍

n this case, the state root is generally not available and the terminated state account is created without setting the sidechain state root. Instead, the account stores the mainchain state root at the time of termination.
A state recovery initialization command will in the future set the sidechain state root by giving an inclusion proof against this mainchain state root.

In the second case, the command contains an inclusion proof against the mainchain state root, proving that another sidechain is either terminated (the status in the chain account is set to 'terminated'), or the liveness condition is violated (as the timestamp of the last certificate is too old).
Therefore, when a CCM reaches a receiving chain that has been terminated, a sidechain terminated message is created and sent back to the sending chain carrying the `stateRoot` of the terminated sidechain.
The application of this CCM on the sidechain will effectively initiate the recovery process.

In both cases, the command sets the sidechain state root in the terminated state account, so that state recovery commands can be issued.
The relevant schemas can be seen below.

> The command ID is `COMMAND_ID_STATE_RECOVERY_INITIALIZATION` and the state recovery initialization command schema can be seen below:

Author: Thanks Sergey! I have updated it 👍

> .State Recovery Initialization Command Schema

Author: OK, as requested I have removed the old schema for the State Recovery Initialization Command and replaced it with the schema in the LIP 54 proposal link above 👍

[%collapsible]
====
[source,js]
----
stateRecoveryInitializationParams = {
    "type": "object",
    "required": [
        "chainID",
        "sidechainChainAccount",
        "bitmap",
        "siblingHashes"
    ],
    "properties": {
        "chainID": {
            "dataType": "uint32",
            "fieldNumber": 1
        },
        "sidechainChainAccount": {
            "dataType": "bytes",
            "fieldNumber": 2
        },
        "bitmap": {
            "dataType": "bytes",
            "fieldNumber": 3
        },
        "siblingHashes": {
            "type": "array",
            "items": {
                "dataType": "bytes"
            },
            "fieldNumber": 4
        }
    }
}
----
====

> Assumming that `trs` is a transaction with module ID `MODULE_ID_INTEROPERABILITY`, and the command ID `COMMAND_ID_STATE_RECOVERY_INITIALIZATION` is to be verified, this can be seen as shown in the following validation schema below:

Author: Thanks, now done 👍

> Assumming that `trs` is a transaction with module ID `MODULE_ID_INTEROPERABILITY`, and the command ID `COMMAND_ID_STATE_RECOVERY_INITIALIZATION` is to be verified, this can be seen as shown in the following validation schema below:

Author: Thanks, these are now updated 👍

> .State Recovery Initialization Command Validation

Author: OK, I have updated all the schemas now in this section.

if terminatedStateAccount(trs.params.chainID) exists and terminatedStateAccount(trs.params.chainID).initialized == True:
    return False

let deserializedInteropAccount be the deserialization of trs.params.sidechainChainAccount
# The commands fails if the sidechain is not terminated on the mainchain but the sidechain already violated the liveness requirement
if (deserializedInteropAccount.status != CHAIN_TERMINATED
    and chainAccount(MAINCHAIN_ID).lastCertificate.timestamp - deserializedInteropAccount.lastCertificate.timestamp <= LIVENESS_LIMIT):
    return False

interopAccKey = uint32be(MODULE_ID_INTEROPERABILITY) || STORE_PREFIX_CHAIN_DATA || uint32be(trs.params.chainID)

query = { key: interopAccKey,
          value: SHA-256(trs.params.sidechainChainAccount),
          bitmap: trs.params.bitmap }

proofOfInclusion = { siblingHashes: trs.params.siblingHashes, queries : [query] }

if terminatedStateAccount(trs.params.chainID) exists:
    return SMTVerify([interopAccKey], proofOfInclusion, terminatedStateAccount(trs.params.chainID).mainchainStateRoot)
else:
    return SMTVerify([interopAccKey], proofOfInclusion, chainAccount(MAINCHAIN_ID).lastCertificate.stateRoot)
----

====

> To continue and process a transaction `trs` with the module ID `MODULE_ID_INTEROPERABILITY` and the command `ID COMMAND_ID_STATE_RECOVERY_INITIALIZATION`, this implies the following logic as displayed below in the state recovery initialization command execution:

Author: OK, so they should now be:  'MODULE_INTEROPERABILITY' and 'COMMAND_STATE_RECOVERY_INITIALIZATION' 👍

> To continue and process a transaction `trs` with the module ID `MODULE_ID_INTEROPERABILITY` and the command `ID COMMAND_ID_STATE_RECOVERY_INITIALIZATION`, this implies the following logic as displayed below in the state recovery initialization command execution:

Author: Done 👍

> .State Recovery Initialization Command Execution

Author: OK, Well I have tried where possible to mirror the LIP from a format and flow perspective.

[%collapsible]
====
[source,js]
----
let sidechainAccount be the deserialization of trs.params.sidechainChainAccount
if terminatedStateAccount(trs.params.chainID) exists:
terminatedStateAccount(trs.params.chainID).stateRoot = sidechainAccount.lastCertificate.stateRoot
terminatedStateAccount(trs.params.chainID).mainchainStateRoot = EMPTY_BYTES
terminatedStateAccount(trs.params.chainID).initialized = True
else:
createTerminatedStateAccount(trs.params.chainID, sidechainAccount.lastCertificate.stateRoot)
----

====

> For further information, including the `createTerminatedStateAccount` function can be found in {url_lip}[LIP 0045^].

Author: I think you mean the link? It should be fine now 👍

==== State Recovery Command

This command is used to recover a certain state (for example, fungible and non-fungible tokens), from a sidechain that has been terminated.
The user proves the existence of an entry in the terminated state substore with an inclusion proof against the state root stored in the sidechain terminated account.
The recovery of the state is then handled by the relevant module (for example, the token module would refund the user), and the state root is updated.

The proof is validated against the current state root stored in the terminated chain account.
The recovery of the state is then handled by the relevant module (for example the token module would refund the user).

The command ID is `COMMAND_ID_STATE_RECOVERY` and the state recovery command schema, verification, and execution can be seen below:

> .State Recovery Command Schema

Author: OK, I have added in the updated schema from the link above 👍

.State Recovery Command Execution
[%collapsible]
====
[source,js]
----
let storeQueries be an empty array

for each entry in trs.params.storeEntries:
    # The recover function corresponding to the module ID applies the recovery logic
    route processing logic to the module given by trs.params.moduleID
    recover(trs.params.chainID, trs.params.moduleID, entry.storePrefix, entry.storeKey, entry.storeValue)
    if the previous call fails:
        execution fails and trs is invalid

    emptyStore = empty bytes # Define an empty store entry
    query = { key: entry.storekey,
              value: SHA-256(emptyStore),
              bitmap: entry.bitmap}
    push query to storeQueries

terminatedStateAccount(trs.params.chainID).stateRoot = SMTCalculateRoot(trs.params.siblingHashes, storeQueries)
----

====

Regarding the recover function, for the verification and application of this command it is assumed that the module given by `trs.params.moduleID` exposes a `recover` function, with the following interface:

[source:js]
----

> recover(terminatedChainID, moduleID, storePrefix, storeKey, storeValue),

Author: OK thanks, now done 👍

----

> ==== Message Recovery Command

Reviewer 2: Maybe create a new section "How to recover messages from the terminated chain outbox" similar to "How to recover information from a sidechain".

Author: OK, I have added a short overview description of the terminated chain outbox in the last section, 'Recovery process workflow'.

Reviewer 2: What I meant is that you have a header of level 4 "Message Recovery Command". The previous header of level 3 is "State recovery from the Sidechain Root", but structurewise this is not correct, as message recovery is not a subtopic of state recovery. 
So I suggest you introduce a new level 3 header "Message recovery" here.

Author: It should be done now 👍

> This command is used to recover pending messages from the outbox of a terminated sidechain.

Reviewer 2: This exact formulation was a bit confusing to me, so we have updated it in the LIP to:
"This mechanism allows to recover any CCM pending in the sidechain channel outbox. That is, those CCMs sent from the sidechain that have not been included in the (terminated) receiving chain yet."

Author: OK thanks Sergey. I have now modified this according to your statement above 👍

The user proves with an inclusion proof validated against the outbox root stored in the terminated outbox account, that the message is in the terminated sidechain outbox.

> The message nonce is compared to the size of the inbox (this is also stored in the terminated outbox account), to check that the message had not been processed yet.

Reviewer 2: The size of the inbox is not stored in the terminated outbox account, it is taken from the channel stored on the terminated sidechain.

The recovered message is then bounced back to the original sending chain or processed directly if the sending chain is the mainchain.

The command ID is `COMMAND_ID_MESSAGE_RECOVERY` and the state recovery command schema can be seen below:

> .Message Recovery Command Schema

Reviewer 2: See the current version of the schemas and the logic here: https://github.com/LiskHQ/lips/blob/187744ada9ce73b1543970b123f69b61b41c64e2/proposals/lip-0054.md#message-recovery-command

    elif ownChainAccount.ID == MAINCHAIN_ID:
        # Only send messages to active and live chains
        if (chainAccount(ccm.receivingChainID) exists
            and chainAccount(ccm.receivingChainID).status == CHAIN_ACTIVE
            and isLive(ccm.receivingChainID):
            addToOutbox(ccm.receivingChainID, ccm)
----

====

==== Message Recovery Initialization Command

Once a terminated state account has been created on the mainchain, users can send this command to initialize the corresponding terminated outbox account.
This command contains the channel account of the mainchain stored in the state of the terminated sidechain and an inclusion proof for it against the state root stored in the corresponding terminated state account.
The mainchain channel data is needed to acquire the number of messages that were processed in the terminated sidechain (corresponding to the inbox size of the channel).

This command sets the outbox root and the inbox size in the terminated outbox account, so that message recovery commands can be issued.

The command ID is `COMMAND_ID_MESSAGE_RECOVERY_INITIALIZATION` and the message recovery initialization command schema can be seen below:

> .Message Recovery Initialization Command Schema

Author: OK, thanks very much Sergey. I have slightly reworded this section and added in the updated schemas from your link above. Can you please check this together with the schemas. 👍

          bitmap: trs.params.bitmap }

proofOfInclusion = { siblingHashes: trs.params.siblingHashes, queries : [query]}

return SMTVerify([channelKey], proofOfInclusion, terminatedStateAccount(trs.params.chainID).stateRoot)
----

====

Finally, processing a transaction `trs` with the module ID `MODULE_ID_INTEROPERABILITY`, and the command ID `COMMAND_ID_MESSAGE_RECOVERY_INITIALIZATION` implies the following logic as can be seen below:

.Message Recovery Initialization Command Execution Schema
[%collapsible]
====
[source,js]
----
let deserializedChannel be the deserialization of trs.params.channel
createTerminatedOutboxAccount(
  trs.params.chainID,
  channel(trs.params.chainID).outbox.root,
  channel(trs.params.chainID).outbox.size,
  deserializedChannel.inbox.size
  )
----

====

== Recovery Process Workflow

As the relevant commands have been described above the process to initiate a recovery is covered in this section.
As can be seen here, the process is slightly different on the mainchain and the sidechain.

==== Mainchain

Firstly regarding the mainchain, the process required is as follows, and is depicted in the illustration in figure 3 below:

> 1. The sidechain is terminated for violating the liveness requirement or the token-module protocol.

Reviewer 2: The sidechain is terminated by submitting a Liveness Termination Command.

Author: Now updated 👍

a. A *terminated state account* containing the last certified state root of the sidechain is created in the state.

b. A User can start recovering assets with a *state recovery command* as covered above.

2. A *message recovery initialization command* is then sent.
This command contains the inbox size of the mainchain stored on the terminated sidechain, which indicates how many cross-chain messages were processed.

a. A *terminated outbox account* is created in the state.
This contains the outbox root of the terminated sidechain channel, and the inbox size provided with the message recovery initialization command.

b. A user can now recover pending messages with a *message recovery command*.

> image:understand-blockchain/interop/mainchain-recovery.png[, align="center" ,500]

Reviewer 2: The link seems to be broken, but the picture itself is good.

Author: The link should now be fixed 👍

Figure 3: Recovery process from the mainchain.

==== Sidechain

> On a sidechain, the process flow can occur in 3 scenarios which are described below and are also depicted in the corresponding 3 diagrams.

Reviewer 2: You don't mention anything about message recovery in this subsection. Basically it will be the same as on the mainchain.

Author: As discussed earlier in Slack, I have now added a sidechain message recovery sub-section 👍

This message contains the last certified state root (on the mainchain), of the terminated sidechain.

a. A *terminated state account* containing the sidechain’s last certified state root is created in the state.
However, it is important to note that in this case, the terminated sidechain state root is set immediately, and there is no need for a state recovery initialization command.

b. A user can recover assets native to the sidechain with a *state recovery command*.

3. The sidechain is terminated on the mainchain.

a. A *state recovery initialization command* is sent.
This command contains the last certified state root (on the mainchain), of the terminated sidechain and an inclusion proof against the last certified mainchain state root, proving the termination.

b. A *terminated state account* containing the sidechain’s last certified state root is created in the state.

c. A user can recover assets that are native to the sidechain with a *state recovery command*.

As mentioned above these 3 steps are depicted graphically below in figure 4.

> image:understand-blockchain/interop/sidechain-recovery.png[, align="center" ,500]

Reviewer 2: The same about pictures here

Author: Very strange... the path is correct, and the pictures are all viewable in my Webstorm :thinking_face:

When a first valid cross-chain update command from this sidechain is processed, the sidechain status is changed to active, making it active in the ecosystem. Now it is possible to send CCMs to the sidechain and the liveness condition is enforced.

=== Sidechain Structure on the Mainchain

As discussed, once a new sidechain is registered on the mainchain by using the registration command, this results in a sidechain account being created in the Lisk mainchain state.
// TO DO Therefore, this implies that the corresponding key-value entries for the sidechain are added to the interoperability module store as shown below in Figure 1.
// TO DO (LIP 0045 - properties of the interoperability module store to refer to this...)

The values of these entries are initialized, and the `name`, `networkID`, and `initValidators`, properties are computed from the sidechain registration command.

This <<Sidechain Registration Command>> assigns a unique integer as the chain ID that identifies the sidechain in the ecosystem.
Furthermore, new data structures are inserted for the sidechain in the Lisk mainchain state.

> This results in a new entry being created in the five different substores of the interoperability module store listed below, and can also be seen in the diagram below in figure 1.

Author: Yes indeed, I meant figure 2... it is now updated, thanks 👍

image:understand-blockchain/interop/interoperability-module-store.png[ align="center" ,800]

Figure 2: The Interoperability module store.

Each box represents a substore, where we indicate the `storeKey --> storeValue relation`.
For the Lisk mainchain, the 'own chain' substore exists by default in the state whereas there is one entry per registered sidechain for five other substores (outbox root, chain data, channel data, chain validators, registered names), created by a sidechain registration command.
For sidechains, the 'own chain' and one entry for the mainchain account for four other substores (outbox root, chain data, channel data, and chain validators) are created by the <<Mainchain Registration Command>>.

=== Sidechain Registration Command

The sidechain registration command is used to register a sidechain on the Lisk mainchain.
Once this command is processed, a new account for the sidechain is created in the mainchain state under the interoperability store.
The account is initialized with an empty inbox and outbox, while the sidechain name and the initial validators set are given in the command parameters.

> The network ID is calculated from the address of the command sender and the genesis block ID, also given in the command parameters.

Author: Thanks, it is now removed 👍

In order to connect a new sidechain to the ecosystem, the sidechain registration command contains the following parameters:

* *name*

The `name` property sets the name of the sidechain as a string of characters, and has to be unique within the Lisk ecosystem, and contain only characters from the set *a-z0-9!@$&_.*

* *initValidators*

The `initValidators` property defines the set of eligible xref:{url_bls}[BLS public keys] with their respective xref:{url_bft_weights}[BFT weights] required to sign the first certificate from the sidechain.

* *certificateThreshold*

The `certificateThreshold` property is an integer setting the minimum signatures weight required for the first sidechain certificate to be valid.

* *sidechainRegistrationFee*

The `sidechainRegistrationFee` property accounts for the extra fee required to register the sidechain. It should be set to the value of the `REGISTRATION_FEE` constant.

> * *Chain ID*

Author: OK, I have updated all instances of chain ID, so they are now "chainID"

> * *Chain ID*

Reviewer: But I would still use lower case ("chainID" instead of "ChainID") in line 133...for consistency with the list :)

> * *Chain ID*

Author: Sounds good...It should now be updated 👍

> The `Chain ID` is responsible for uniquely identifying a chain in the Lisk ecosystem.

Author: Done 👍

Just as addresses are used for the identification of user accounts, the chain ID has a similar purpose.

> When processing the sidechain registration command, the chain ID for a sidechain is deterministically computed.

Author: Updated accordingly 👍

Specifically, the chain ID of a new sidechain is assigned as an incremental integer similar to transaction nonces.
The format of chain IDs aims to provide an efficient and compact way to uniquely identify chains in the ecosystem.

> Furthermore, an additional advantage, is that it is easy to remember the integer assigned as the chain ID for a users favorite blockchain application.

Author: I presume you mean all 3 lines 174 to 176 above, so they have now all been removed 👍

The chain ID is a 4-byte constant set in the chain configuration.
Chain IDs serve the following two purposes:

*(1)* Chain IDs are prepended to the input of the signing function of every transaction, block, or message of the chain to avoid transaction replays between different chains in the ecosystem.

*(2)* Chain IDs uniquely identify a chain in the Lisk ecosystem.

> Specifically, the Interoperability module, it serves a similar purpose for chains as addresses do for user accounts, as it is used to identify the chain account in the Interoperability module store.

Author: OK, the last sentence above has now be removed 👍

Therefore, the sidechain has to change the chain ID with a hardfork and resubmit the sidechain registration command with a new value.
The chain identifiers are of a value of 4 bytes, and dependent on the network on which the chain is running, the first byte must always be set to the correct value.

An example can be seen below in the following table depicting the chain-identifiers prefixes currently specified.
The first byte is set to  `CHAIN_ID_PREFIX_MAINNET` for chains running in the mainnet network and to `CHAIN_ID_PREFIX_TESTNET` for chains running in the testnet network.
The other 3 bytes must be uniquely chosen for the respective blockchain, hence, no other blockchain created with the Lisk SDK should use the same 3 bytes.

[cols="2,1,2,4"]
|===
|Name|Type|Value|Description
|`CHAIN_ID_PREFIX_MAINNET`|bytes|0x00|Chain-identifier prefix for mainnet blockchains.
|`CHAIN_ID_PREFIX_TESTNET`|bytes|0x01|Chain-identifier prefix for testnet blockchains.
|===

The chain ID is known to the mainchain as soon as the sidechain is registered, therefore it can validate cross-chain update commands coming from the sidechain without any further context.

> === Sidechain Network Identifier

Author: OK, the 'Sidechain Network Identifier' section has now been removed.

It is then possible to start the sidechain at this point without being registered on the mainchain.
In addition, the network ID is known to the mainchain as soon as the sidechain is registered, therefore it can validate cross-chain update commands coming from the sidechain without any further context.

Note that in the case of a sidechain undergoing a community hard fork, one of the competing forks will need to register their fork of the sidechain again on the Lisk Mainchain.
Regardless of the genesis block ID set by this sidechain fork, it has to be registered from a different user account which implies that the chain ID and network ID of the forked chain will be different from the original one.

=== Mainchain Registration on a Sidechain

Once the sidechain has been registered on the mainchain, a similar registration process should happen in the sidechain before the interoperable channel is opened between the two chains.
This is performed by submitting a transaction with the mainchain registration command in the sidechain, which implies the creation of a mainchain account in the sidechain state associated with the Lisk mainchain, and other structures needed for interoperability.

> This mainchain account has a similar structure as the one depicted in Figure 1 above.

Author: Yes indeed, many thanks and now updated to figure 2 👍

> By protocol, the chain ID of the mainchain is a constant equal to `CHAIN_ID_MAINCHAIN` in the ecosystem.

Author: Thanks, now updated as recommended 👍

This registration process always has to occur after the sidechain registration on the mainchain, since the sidechain has no prior knowledge of the current mainchain validators or its name.
Similar to the sidechain registration case, the mainchain account status will not change to active until a valid cross-chain update command from the mainchain containing a valid registration xref:{url_ccm}[CCM] is processed.

=== Mainchain Registration Command

The mainchain registration command is used to register the Lisk mainchain on a sidechain.
When this command is processed, a new account for the mainchain is created in the sidechain state under the interoperability store.

Certain parameters are set by the mainchain registration command in the sidechain that is related to the interoperability module, and also initialize the corresponding mainchain data structures.

> This command requires the approval of the sidechain validators, as they have to agree on the content of this command and add their aggregated signatures accordingly.

Reviewer: ```suggestion
This command requires the approval of the sidechain validators, as they have to agree on the content of this command and add their aggregate signatures accordingly.
```

It is of key importance that the sidechain validators ensure that they are signing the registration command with the correct information, otherwise the sidechain interoperable functionality may be unusable.

There is no minimum fee required for this command, as it should only be submitted once in a sidechain and approved by a majority of validators.
Taking this into account, a transaction with this command should be treated differently in terms of priority in case it is included in a sidechain node's transaction pool.

> Hence, it is recommended that once the transaction is correctly signed by the validators and ready to be submitted, a validator simply includes it in its next forged block, without including it in the transaction pool.

Reviewer: This was important for the Rationale section of the LIP. But I guess readers of the documentation are not interested in this.

Author: OK, fair point...It has now been removed 👍

The command has the following parameters:

* *ownChainID*

The chain ID is set on the mainchain after processing the corresponding sidechain registration command.

* *ownName*

The `ownName` property sets the name of the sidechain in its own state according to the name given in the mainchain.

* *mainchainValidators*

This is similar to the `initValidators` property in the sidechain registration command, it defines the set of mainchain validators with their respective xref:{url_bft_weights}[BFT weight] expected to sign the first certificate from the mainchain.

* *signature*

> The `signature` property is an aggregated signature of the sidechain validators.

Reviewer: ```suggestion
The `signature` property is an aggregate signature of the sidechain validators.
```

It ensures that the sidechain validators agree on registering the mainchain in the sidechain.

* *aggregationBits*

> The `aggregationBits` property is a bit vector used to validate the aggregated signature.

Reviewer: ```suggestion
The `aggregationBits` property is a bit vector used to validate the aggregate signature.
```

        },
        "sidechainRegistrationFee": {
            "type": "uint64",
            "fieldNumber": 5
        }
    }
}
----
.Verification

[source,js]
----
def verify(trs: Transaction) -> None:
    # The name property has to contain only characters from the set [a-z0-9!@$&_.].
    if not re.match(r"^[a-z0-9!@$&_.]+$", trs.params.name):
        raise Exception("Invalid name property. It should contain only characters from the set [a-z0-9!@$&_.].")
    # The name property has to be unique with respect to the set of already registered sidechain names.
    if there exists an entry in the registered names substore with store key equal to trs.params.name.encode("utf-8"):
        raise Exception("Name already registered.")

    # Chain ID has to be unique with respect to the set of already registered sidechains.
    if there exists an entry in the chain data substore with store key equal to trs.params.chainID:
        raise Exception("Chain ID already registered.")

    # Check that the first byte of the chainID, indication the network, matches.
    if trs.params.chainID[0] != CHAIN_ID_MAINCHAIN[0]:
        raise Exception("Chain ID does not match the mainchain network.")

> 

Reviewer: ```suggestion
    # Chain ID cannot be the mainchain chain ID.
    if trs.params.chainID == CHAIN_ID_MAINCHAIN:
        raise Exception("Chain ID cannot be the mainchain chain ID.")
```

    # Create validators account.
    sidechainValidators = {
        "activeValidators": trs.params.initValidators,
        "certificateThreshold": trs.params.certificateThreshold
    }
    create an entry in the chain validators data substore with
        storeKey = chainID
        storeValue = encode(validatorsSchema, sidechainValidators)

    # Create outbox root entry.
    create an entry in the outbox root substore with
        storeKey = chainID
        storeValue = encode(outboxRootSchema, {"root": sidechainChannel.outbox.root})

    # Create registered names entry.
    create an entry in the registered names substore with
        storeKey = trs.params.name
        storeValue = encode(chainIDSchema, {"chainID": chainID})

    # Burn the registration fee.
    Token.burn(senderAddress, TOKEN_ID_LSK_MAINCHAIN, REGISTRATION_FEE)

    # Emit chain account updated event.
    emitEvent(
        module = MODULE_NAME_INTEROPERABILITY,
        name = EVENT_NAME_CHAIN_ACCOUNT_UPDATED,
        data = sidechainAccount,
        topics = [chainID]
    )

    # Send registration CCM to the sidechain.
    registrationCCMParams = {

>         "chainID": chainID,

Reviewer: ```suggestion
```

        "name": trs.params.name,
        "messageFeeTokenID": sidechainChannel.messageFeeTokenID
    }

    ccm = {
        "nonce": ownChainAccount.nonce,
        "module": MODULE_NAME_INTEROPERABILITY,
        "crossChainCommand": CROSS_CHAIN_COMMAND_REGISTRATION,
        "sendingChainID": ownChainAccount.chainID,
        "receivingChainID": chainID,
        "fee": 0,
        "status": CCM_STATUS_CODE_OK,
        "params": encode(registrationCCMParamsSchema, registrationCCMParams) # registrationCCMParamsSchema is defined in LIP0049
    }

    addToOutbox(chainID, ccm)
    ownChainAccount.nonce += 1

>     # Emit CCM Processed Event.

Reviewer: ```suggestion
    # Emit CCM Sent Event.
```

    ccmID = sha256(encode(crossChainMessageSchema, ccm))
    emitEvent(
        module = MODULE_NAME_INTEROPERABILITY,

>         name = EVENT_NAME_CCM_SEND_SUCCESS,

Reviewer: ```suggestion
        name = EVENT_NAME_CCM_SENT_SUCCESS,
```

        data = {"ccmID": ccmID},
        topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
    )
----
====

=== Mainchain Registration Transaction Commands

Secondly, the transactions executing the mainchain registration command and the required parameters, verification, and execution are listed below:

* `module = MODULE_NAME_INTEROPERABILITY`
* `command = COMMAND_MAINCHAIN_REG`

.Mainchain Parameters

> [%collapsible]

Reviewer: It's hard to read this JSON file when it's flat.

Author: It should be better now as the mainchain, verification, and execution schemas can now be viewed separately.

Reviewer: What I meant is there are no indentations used in this JSON which makes it very hard to see the nesting.

Author: Ah yes, sorry. It should be fixed now 👍

        "certificateThreshold": THRESHOLD_MAINCHAIN
    }
    create an entry in the chain validators data substore with
        storeKey = CHAIN_ID_MAINCHAIN
        storeValue = encode(validatorsSchema, mainchainValidators)

    # Create outbox root entry.
    create an entry in the outbox root substore with
        storeKey = CHAIN_ID_MAINCHAIN
        storeValue = encode(outboxRootSchema, {"root": mainchainChannel.outbox.root})

    # Create own chain account.
    ownChainAccount = {
        "name": trs.params.ownName,
        "chainID": trs.params.ownChainID,
        "nonce": 0
    }
    create an entry in the own chain data substore with
        storeKey = EMPTY_BYTES
        storeValue = encode(ownChainAccountSchema, ownChainAccount)

    # Emit chain account updated event.
    emitEvent(
        module = MODULE_NAME_INTEROPERABILITY,
        name = EVENT_NAME_CHAIN_ACCOUNT_UPDATED,
        data = mainchainAccount,
        topics = [CHAIN_ID_MAINCHAIN]
    )

    # Send registration CCM to the mainchain.
    # Notice that we do not use the send function because the channel
    # has not been activated yet.
    registrationCCMParams = {

>         "chainID": CHAIN_ID_MAINCHAIN,

Reviewer: ```suggestion
```

        "name": CHAIN_NAME_MAINCHAIN,
        "messageFeeTokenID": mainchainChannel.messageFeeTokenID
    }

    ccm = {
        "nonce": ownChainAccount.nonce,
        "module": MODULE_NAME_INTEROPERABILITY,
        "crossChainCommand": CROSS_CHAIN_COMMAND_REGISTRATION,
        "sendingChainID": ownChainAccount.chainID,
        "receivingChainID": CHAIN_ID_MAINCHAIN,
        "fee": 0,
        "status": CCM_STATUS_CODE_OK,
        "params": encode(registrationCCMParamsSchema, registrationCCMParams) # registrationCCMParamsSchema is defined in LIP0049
    }

    # We cannot use the send function because the channel is not active yet.
    addToOutbox(CHAIN_ID_MAINCHAIN, ccm)
    ownChainAccount.nonce += 1

>     # Emit CCM Processed Event.

Reviewer: ```suggestion
    # Emit CCM Sent Event.
```

    ccmID = sha256(encode(crossChainMessageSchema, ccm))
    emitEvent(
        module = MODULE_NAME_INTEROPERABILITY,

>         name = EVENT_NAME_CCM_SEND_SUCCESS,

Reviewer: ```suggestion
        name = EVENT_NAME_CCM_SENT_SUCCESS,
```

        terminatedStateAccount(trs.params.chainID).mainchainStateRoot = EMPTY_BYTES
        terminatedStateAccount(trs.params.chainID).initialized = True
    else:
        createTerminatedStateAccount(trs.params.chainID, sidechainAccount.lastCertificate.stateRoot)
----

====
For further information, including the `createTerminatedStateAccount` function can be found in xref:{url_lip}[LIP 0045].

==== State Recovery Command

This command is used to recover a certain state (for example, fungible and non-fungible tokens), from a sidechain that has been terminated.
The user proves the existence of an entry in the terminated state substore with an inclusion proof against the state root stored in the sidechain terminated account.
The recovery of the state is then handled by the relevant module (for example, the token module would refund the user), and the state root is updated.

The proof is validated against the current state root stored in the terminated chain account.
The recovery of the state is then handled by the relevant module (for example the token module would refund the user).

> The module name is `MODULE_NAME_INTEROPERABILITY`, and the command ID is `COMMAND_STATE_RECOVERY` and the state recovery command schema, verification, and execution can be seen below:

Reviewer 2: Here and in some other occurrences you use the notion "command ID". We have refactored this in some of the recent updates, now it is called "command name".
You can just search in the document and replace all "command ID" with "command name".

Author: Done 👍

        "module": {
            "dataType": "bytes",
            "length": MODULE_ID_LENGTH,
            "fieldNumber": 2
        },
        "storeEntries": {
            "type": "array",
            "fieldNumber": 3,
            "items": {
                "type": "object",
                "properties": {
                    "substorePrefix": {
                        "dataType": "bytes",
                        "fieldNumber": 1
                    },
                    "storeKey": {
                        "dataType": "bytes",
                        "fieldNumber": 2
                    },
                    "storeValue": {
                        "dataType": "bytes",
                        "fieldNumber": 3
                    },
                    "bitmap": {
                        "dataType": "bytes",
                        "fieldNumber": 4
                    }
                },
                "required": ["substorePrefix", "storeKey", "storeValue", "bitmap"]
            }
        },
        "siblingHashes": {
            "type": "array",
            "items": {
                "dataType": "bytes"
            },
            "fieldNumber": 4
        }
    }
}
----

====

> Assuming `trs` is a transaction with the module `ID MODULE_ID_INTEROPERABILITY`, and the command ID `COMMAND_STATE_RECOVERY` is to be verified.

Reviewer 2: Here and in some other occurrences you use the notion "module ID". We have refactored this in some of the recent updates, now it is called "module name".
You can just search in the document and replace all "module ID" with "module name" and "`MODULE_ID`" with "`MODULE`".

Author: OK cool...Many thanks, it should now be done... 👍

Then `trs` will be valid if the following logic returns `True`, as can be seen below:

> .State Recovery Command Verification

Reviewer 2: Here and in the State Recovery Command Execution the code snippets need to be updated according to the last version: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0054.md#verification.
Note that the PR that was open is merged now.

Author: OK, these have been updated... In addition, the State Recovery Command section has also been updated 👍

messageRecoveryParams = {
    "type": "object",
    "required": ["chainID", "crossChainMessages", "idxs", "siblingHashes"],
    "properties": {
        "chainID": {
            "dataType": "bytes",
            "length": CHAIN_ID_LENGTH,
            "fieldNumber": 1
        },
        "crossChainMessages": {
            "type": "array",
            "items": {
                "dataType": "bytes"
            },
            "fieldNumber": 2
        },
        "idxs": {
            "type": "array",
            "items": {
                "dataType": "uint32"
            },
            "fieldNumber": 3
        },
        "siblingHashes": {
            "type": "array",
            "items": {
                "dataType": "bytes",
                "length": HASH_LENGTH
            },
            "fieldNumber": 4
        }
    }
}
----

====

* `chainID`: The ID of the terminated sidechain identifying the terminated outbox from which messages will be recovered.
* `crossChainMessages`: The cross-chain messages to be recovered.
* `idxs`: The indices of the messages to be recovered.
* `siblingHashes`: The sibling hashes of the inclusion proof of the cross-shain messages in the sidechain outbox.

> .Message Recovery Command Verification Schema

Reviewer 2: This is not a schema, this is just "Message recovery command verification". Schema is a JSON file that describes a particular data type (for example, parameters for a command).
The same concerns "Message Recovery Command Execution" below, as well as Message Recovery Initialization and Liveness Termination commands.

Author: OK, Iall instances of 'schema' when describing the commands should now be removed 👍

        raise Exception("Chain is not terminated.")

    # The commands fails if there exist already a terminated outbox account
    if terminatedOutboxAccount(trs.params.chainID) exists:
        raise Exception("Terminated outbox account already exists.")

    queryKey = STORE_PREFIX_INTEROPERABILITY + SUBSTORE_PREFIX_CHANNEL_DATA + sha256(CHAIN_ID_MAINCHAIN)

    query = {
        key: queryKey,
        value: sha256(trs.params.channel),
        bitmap: trs.params.bitmap
    }

    proofOfInclusion = { siblingHashes: trs.params.siblingHashes, queries : [query] }

    if SMTVerify([queryKey], proofOfInclusion, terminatedStateAccount(trs.params.chainID).stateRoot) == False:
        raise Exception("Message recovery initialization proof of inclusion is not valid.")
----

====

.Message Recovery Initialization Command Execution Schema
[%collapsible]
====
[source,js]
----
def execute(trs: Transaction) -> None:
    partnerChannel = decode(channelSchema, trs.params.channel)
    createTerminatedOutboxAccount(
        trs.params.chainID,
        channel(trs.params.chainID).outbox.root,
        channel(trs.params.chainID).outbox.size,
        partnerChannel.inbox.size
----

====

> === Liveness Termination Command

Reviewer 2: I think, this should be level 4 header, not level 3.

Author: Yes, you are right, it should be fixed now 👍

This command is used to terminate a sidechain that violated the liveness condition.
As such, it can only be submitted on the mainchain.
The required parameters, verification, and execution are listed below:

* `module = MODULE_NAME_INTEROPERABILITY`
* `command = COMMAND_LIVENESS_TERMINATION`

.Liveness Termination Command Schema
[%collapsible]
====
[source,js]
----
livenessTerminationParams = {
    "type": "object",
    "required": ["chainID"],
    "properties": {
        "chainID": {
            "dataType": "bytes",
            "length": CHAIN_ID_LENGTH,
            "fieldNumber": 1
        }
}
----

====

> * `The chainID` is the ID of the respective sidechan to be terminated.

Reviewer 2: Typo: The `chainID`.

Author: well spotted! 👍

As mentioned above the steps from 1 to 3 are depicted graphically below in figure 4.

===== Message Recovery from the Sidechain Outbox

This message recovery mechanism is able to recover any CCM that is pending in the sidechain outbox.
Specifically, those CCMs that have not yet been included in the receiving sidechain.
Hence, this includes all the CCMs whose indices are larger than the last message index that the receiving sidechain reported to have included in its inbox on the mainchain.
Any user can submit a message recovery command to recover several CCMs simultaneously.
When the command is processed, the recovery is dependent on whether the sending chain is the mainchain or a sidechain.

However, it is important to note that users are not guaranteed to recover their CCMs in every situation.
Certain state information of the terminated sidechain might have been modified before the termination occurred, and this would make the recovered CCM application fail.
For example, in the case whereby escrowed LSK exists in the sidechain account on the mainchain, which could have been subtracted by prior malicious behavior in the terminated sidechain.

> image::understand-blockchain/interop/sidechain-recovery1-v2.png[, align="center" ,500]
> 
> image::understand-blockchain/interop/sidechain-recovery2_v2.png[, align="center" ,500]
> .Recovery process from the sidechain.
> image::understand-blockchain/interop/sidechain-recovery3-v2.png[, align="center" ,500]

Reviewer 3: ```suggestion
.Recovery process from the sidechain (Step 1)
image::understand-blockchain/interop/sidechain-recovery1-v2.png[, align="center" ,500]
.Recovery process from the sidechain (Step 2)
image::understand-blockchain/interop/sidechain-recovery2_v2.png[, align="center" ,500]
.Recovery process from the sidechain (Step 3)
image::understand-blockchain/interop/sidechain-recovery3-v2.png[, align="center" ,500]
```

:url_bls: pages/glossary.adoc#bls
:url_bft_weights: pages/glossary.adoc#bft-weight
:url_outbox_rootwitness: understand-blockchain/interoperability/communication.adoc
:url_sdk_token_module: {docs_sdk}modules/token-module.adoc
:url_custom_module: understand-blockchain/sdk/modules-commands
:url_nft_module: tutorial/nft.adoc#the-nft-module
:url_lip: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0045.md

//:url_how_blockchain_works: intro/how-blockchain-works.adoc
//:url_transactions: understand-blockchain/lisk-protocol/transactions.adoc

== Overview - What is a Sidechain?

A sidechain is a separate blockchain next to the mainchain, where the inclusion of a transaction in one of the blockchains can have an effect in the other blockchain.
These effects on the other blockchain can facilitate, for example, moving digital assets between the mainchain and the sidechain.
Having a mainchain and some sidechains in parallel results in a higher overall transaction throughput.

== Sidechains in Lisk

Within the Lisk network, a transaction that can have an effect on another blockchain is called a xref:{url_cct}[cross-chain transaction].

> Such a transaction emits one or more xref:{url_ccm}[cross-chain messages] on the sending chain. The emitted cross-chain messages on the sending chain are then collected by a relayer and transferred to the sending chain via a so-called xref:{url_ccu}[cross-chain updated transaction] (CCU).

Author: Now done 👍

> Such a transaction emits one or more xref:{url_ccm}[cross-chain messages] on the sending chain. The emitted cross-chain messages on the sending chain are then collected by a relayer and transferred to the sending chain via a so-called xref:{url_ccu}[cross-chain updated transaction] (CCU).

Author: Good point...now updated accordingly 👍

> Such a transaction emits one or more xref:{url_ccm}[cross-chain messages] on the sending chain. The emitted cross-chain messages on the sending chain are then collected by a relayer and transferred to the sending chain via a so-called xref:{url_ccu}[cross-chain updated transaction] (CCU).

Author: Yes, it should...now corrected 👍

When the CCU is included in the receiving chain, the cross-chain messages induce some state changes.

> This is how a transaction from a sidechain (the mainchain) can have an effect on the mainchain (a sidechain).

Author: Good point, now reworded to simplify it 👍

Then, users can recover tokens from the terminated sidechain with a <<State Recovery Command>> or proceed with the <<Message Recovery Initialization Command>> to kick off the recovery of pending cross-chain messages.
The sidechain termination and recovery process is covered later on this page.

== How to register a sidechain

As mentioned in the previous paragraph to enable a sidechain to operate, it first has to be registered in the Lisk mainchain via a sidechain registration command.
This requires that the sidechain registration command is processed, which will then result in the sidechain account status being set to registered.
However, at this point, the channel is not active, so no cross-chain messages (CCMs) can be sent yet.
Hence, a sidechain account that is already associated with a name and a unique chain ID, exists in the mainchain state.
Therefore, each new sidechain occupies a certain namespace in the ecosystem.

=== Sidechain Registration Process

A transaction with a <<Sidechain Registration Command>> can be sent by any user account in the Lisk Mainchain that possesses adequate funds to pay the required fee.

> As described in the Lifecycle of a Sidechain section above, once the Sidechains registration process has been completed, it does not need to be activated on the mainchain, and will remain inactive until the first valid CCU containing some cross-chain messages has been sent and processed.

Author: Done 👍

For sidechains, the 'own chain' and one entry for the mainchain account for four other substores (outbox root, chain data, channel data, and chain validators) are created by the <<Mainchain Registration Command>>.

=== Sidechain Registration Command

The sidechain registration command is used to register a sidechain on the Lisk mainchain.
Once this command is processed, a new account for the sidechain is created in the mainchain state under the interoperability store.
The account is initialized with an empty inbox and outbox, while the sidechain name, chain ID, and the initial validators set are given in the command parameters.

In order to connect a new sidechain to the ecosystem, the sidechain registration command contains the following parameters:

* *name*

The `name` property sets the name of the sidechain as a string of characters, and has to be unique within the Lisk ecosystem, and contains only characters from the set *a-z0-9!@$&_.*

* *initValidators*

> The `initValidators` property defines the set of eligible xref:{url_bls}[BLS public keys] with their respective xref:{url_bft_weights}[BFT weights] required to sign the first certificate from the sidechain.

Author: These entries have been removed from the Glossary, so I have removed the corresponding links 👍

When the CCU is included in the receiving chain, the cross-chain messages induce some state changes.
This is how a transaction from a sidechain (the mainchain) can have an effect on the mainchain (a sidechain).

The main purpose of deploying sidechains in the Lisk ecosystem is to increase scalability: The more sidechains that exist in the ecosystem means that more transactions that can be processed per second.
It is intended that each blockchain serves one application, that means, for instance, there could be a sidechain dedicated to a decentralized exchange application.

== Life Cycle of a Sidechain

The life cycle of a sidechain can be broken down into 3 parts, corresponding to the following 3 values of the account status property: ''registered'', ''active'', and ''terminated''.

Firstly, a sidechain registers on the mainchain with a *sidechain registration command*.
This command creates the sidechain account on the mainchain with the initial status set to 'registered'.
Thereafter, the mainchain account is similarly created on a sidechain with a *mainchain registration command*.

> After a sidechain has been registered on the mainchain, it cannot receive any cross-chain message and does not need to follow the liveness rule, until the first sidechain CCU has been included in the mainchain.

Author: No worries...now updated 👍

|`CHAIN_ID_PREFIX_TESTNET`|bytes|0x01|Chain-identifier prefix for testnet blockchains.
|===

The chainID is known to the mainchain as soon as the sidechain is registered, therefore it can validate cross-chain update commands coming from the sidechain without any further context.

=== Mainchain Registration on a Sidechain

Once the sidechain has been registered on the mainchain, a similar registration process should happen in the sidechain before the interoperable channel is opened between the two chains.
This is performed by submitting a transaction with the mainchain registration command in the sidechain, which implies the creation of a mainchain account in the sidechain state associated with the Lisk mainchain, and other structures needed for interoperability.
This mainchain account has a similar structure as the one depicted in Figure 2 above.
By protocol, the chainID of the mainchain is a constant equal to `0x00000000` in the ecosystem.

> This registration process always has to occur after the sidechain registration on the mainchain, since the sidechain has no prior knowledge of the current mainchain validators or its name.

Reviewer: ```suggestion
This registration process always has to occur after the sidechain registration on the mainchain, since the sidechain has no prior knowledge of its name and must be certain that the correct chain ID has been registered.
```

The mainchain registration command is used to register the Lisk mainchain on a sidechain.
When this command is processed, a new account for the mainchain is created in the sidechain state under the interoperability store.

Certain parameters are set by the mainchain registration command in the sidechain that are related to the interoperability module, and also initialize the corresponding mainchain data structures.

> This command requires the approval of the sidechain validators, as they have to agree on the content of this command and add their aggregate signatures accordingly.

Reviewer: If you want, you could also be more precise: "... as they have to agree on the content of this command and provide their signature accordingly. Based on the individual signatures, an aggregate signature must be added to the transaction."

But I guess it's not needed.

> This command requires the approval of the sidechain validators, as they have to agree on the content of this command and add their aggregate signatures accordingly.

Author: I like it...I have now added it, many thanks 👍

[source,js]
----
mainchainRegParams = {
    "type": "object",
    "required": [
        "ownChainID",
        "ownName",
        "mainchainValidators",
        "signature",
        "aggregationBits"
    ],
    "properties": {
        "ownChainID": {
            "dataType": "uint32",
            "fieldNumber": 1
        },
        "ownName": {
            "dataType": "string",
            "fieldNumber": 2
        },
        "mainchainValidators": {
            "type": "array",
            "fieldNumber": 3,
            "items": {
                "type": "object",
                "required": ["blsKey", "bftWeight"],
                "properties": {
                    "blsKey": {

>                         "dataType": "bytes",

Reviewer: ```suggestion
                        "dataType": "bytes",
                        "length": BLS_PUBLIC_KEY_LENGTH,
```

                        "fieldNumber": 1
                    },
                    "bftWeight": {
                        "dataType": "uint64",
                        "fieldNumber": 2
                    }
                }
            }
        },
        "signature": {

>             "dataType": "bytes",

Reviewer: ```suggestion
            "dataType": "bytes",
            "length": BLS_SIGNATURE_LENGTH,
```

Certain state information of the terminated sidechain might have been modified before the termination occurred, and this would make the recovered CCM application fail.
For example, in the case whereby escrowed LSK exists in the sidechain account on the mainchain, which could have been subtracted by prior malicious behavior in the terminated sidechain.

image::understand-blockchain/interop/sidechain-recovery-v1.png[, align="center" ,500]

image::understand-blockchain/interop/sidechain-recovery2_v1.png[, align="center" ,500]
.Recovery process from the sidechain.

> image::understand-blockchain/interop/sidechain-recovery3-v1.png[, align="center" ,500]

Reviewer 3: The images look nice now, however, captions for each are missing for these images. An image must have a caption. Also, Please add Sidechain Recovery Command stage/step **x**, where x could be 1, 2, 3 inside the image to clarify that this is a step-by-step process.

Author: Good point..now done 👍
